### PR TITLE
Fix checkpoint  already exist

### PR DESCRIPTION
**Pull request by ptsayli**

### CODE REVIEW

File: bridge/setu/processor/checkpoint.go [hunk 1]
```diff
@@ -138,8 +138,17 @@ func (cp *CheckpointProcessor) sendCheckpointToHeimdall(headerBlockStr string) (
 		}
 		start := expectedCheckpointState.newStart
 		end := expectedCheckpointState.newEnd
-		// TODO - add a check to see if this checkpoint has to be proposed or not.
-		// Fetch latest checkpoint from buffer. if expectedCheckpointState.newStart == start, don't send checkpoint
+
+		bufferedCheckpoint, err := util.GetBufferedCheckpoint(cp.cliCtx)
+		if err != nil {
+			cp.Logger.Error("No buffered checkpoint", "error", err)
```
Comment (mankenavenkatesh) [inline bridge/setu/processor/checkpoint.go:144]: It's not an error. Buffer being empty is valid.

File: bridge/setu/processor/checkpoint.go [hunk 2]
```diff
@@ -138,8 +138,17 @@ func (cp *CheckpointProcessor) sendCheckpointToHeimdall(headerBlockStr string) (
...
+		}
+
+		if bufferedCheckpoint != nil && bufferedCheckpoint.StartBlock == start {
+			cp.Logger.Debug("Checkpoint already exits in buffer", "Checkpoint", bufferedCheckpoint.String())
```
Comment (mankenavenkatesh) [inline bridge/setu/processor/checkpoint.go:148]: pls print `startBlock` number for debugging purpose.

File: bridge/setu/processor/checkpoint.go [hunk 3]
```diff
@@ -138,8 +138,17 @@ func (cp *CheckpointProcessor) sendCheckpointToHeimdall(headerBlockStr string) (
...
+			return err
```
Comment (mankenavenkatesh) [inline bridge/setu/processor/checkpoint.go:149]: Don't return error. 
when error is returned, machinery will retry the task.

File: bridge/setu/util/common.go [hunk 4]
```diff
@@ -325,6 +325,25 @@ func GetCheckpointParams(cliCtx cliContext.CLIContext) (*checkpointTypes.Params,
...
 
+// GetBufferedCheckpoint return checkpoint from bueffer
+func GetBufferedCheckpoint(cliCtx cliContext.CLIContext) (*hmtypes.CheckpointBlockHeader, error) {
+	response, err := helper.FetchFromAPI(
+		cliCtx,
+		helper.GetHeimdallServerEndpoint(BufferedCheckpointURL),
+	)
+
+	if err != nil {
+		return nil, err
```
Comment (mankenavenkatesh) [inline bridge/setu/util/common.go:336]: pls add error log here. 
Helps in knowing there is failure with external call.

File: bridge/setu/util/common.go [hunk 5]
```diff
@@ -325,6 +325,25 @@ func GetCheckpointParams(cliCtx cliContext.CLIContext) (*checkpointTypes.Params,
...
+	}
+
+	var blockHeader hmtypes.CheckpointBlockHeader
+	if err := json.Unmarshal(response.Result, &blockHeader); err != err {
+		return nil, err
```
Comment (mankenavenkatesh) [inline bridge/setu/util/common.go:341]: pls add error log.

File: bridge/setu/util/common.go [hunk 6]
```diff
@@ -314,17 +316,40 @@ func GetCheckpointParams(cliCtx cliContext.CLIContext) (*checkpointTypes.Params,
...
+	)
+
+	if err != nil {
+		logger.Debug("Error fetching buffered checkpoint", err)
+		return nil, err
+	}
+
+	var blockHeader hmtypes.CheckpointBlockHeader
+	if err := json.Unmarshal(response.Result, &blockHeader); err != err {
+		logger.Error("Error unmarshalling buffered checkpoint", "url", BufferedCheckpointURL)
```
Comment (jdkanani) [inline bridge/setu/util/common.go:346]:
```suggestion
		logger.Error("Error unmarshalling buffered checkpoint", "url", BufferedCheckpointURL, "err",  err)
```

File: bridge/setu/util/common.go [hunk 7]
```diff
@@ -314,17 +316,40 @@ func GetCheckpointParams(cliCtx cliContext.CLIContext) (*checkpointTypes.Params,
...
+		helper.GetHeimdallServerEndpoint(BufferedCheckpointURL),
+	)
+
+	if err != nil {
+		logger.Debug("Error fetching buffered checkpoint", err)
+		return nil, err
+	}
+
+	var blockHeader hmtypes.CheckpointBlockHeader
+	if err := json.Unmarshal(response.Result, &blockHeader); err != err {
```
Comment (jdkanani) [inline bridge/setu/util/common.go:345]:
```suggestion
	if err := json.Unmarshal(response.Result, &blockHeader); err != nil {
```

File: bridge/setu/util/common.go [hunk 8]
```diff
@@ -314,17 +316,40 @@ func GetCheckpointParams(cliCtx cliContext.CLIContext) (*checkpointTypes.Params,
...
 
+// GetBufferedCheckpoint return checkpoint from bueffer
+func GetBufferedCheckpoint(cliCtx cliContext.CLIContext) (*hmtypes.CheckpointBlockHeader, error) {
+	response, err := helper.FetchFromAPI(
+		cliCtx,
+		helper.GetHeimdallServerEndpoint(BufferedCheckpointURL),
+	)
+
+	if err != nil {
+		logger.Debug("Error fetching buffered checkpoint", err)
```
Comment (jdkanani) [inline bridge/setu/util/common.go:340]:
```suggestion
		logger.Debug("Error fetching buffered checkpoint", "err", err)
```

File: bridge/setu/util/common.go [hunk 9]
```diff
@@ -295,11 +295,13 @@ func GetChainmanagerParams(cliCtx cliContext.CLIContext) (*chainManagerTypes.Par
 	)
 
 	if err != nil {
+		logger.Error("Error fetching chainmanager params", err)
 		return nil, err
 	}
 
 	var params chainManagerTypes.Params
 	if err := json.Unmarshal(response.Result, &params); err != nil {
+		logger.Error("Error unmarshalling chainmanager params", "url", ChainManagerParamsURL)
```
Comment (jdkanani) [inline bridge/setu/util/common.go:304]:
```suggestion
		logger.Error("Error unmarshalling chainmanager params", "url", ChainManagerParamsURL, "err", err)
```

File: bridge/setu/util/common.go [hunk 10]
```diff
@@ -314,17 +316,40 @@ func GetCheckpointParams(cliCtx cliContext.CLIContext) (*checkpointTypes.Params,
 	)
 
 	if err != nil {
+		logger.Error("Error fetching Checkpoint params", err)
```
Comment (jdkanani) [inline bridge/setu/util/common.go:319]:
```suggestion
		logger.Error("Error fetching Checkpoint params", "err", err)
```

File: bridge/setu/util/common.go [hunk 11]
```diff
@@ -295,11 +295,13 @@ func GetChainmanagerParams(cliCtx cliContext.CLIContext) (*chainManagerTypes.Par
 	)
 
 	if err != nil {
+		logger.Error("Error fetching chainmanager params", err)
```
Comment (jdkanani) [inline bridge/setu/util/common.go:298]:
```suggestion
		logger.Error("Error fetching chainmanager params", "err", err)
```

**Pull request merged**